File: scripts/rewrite_scorer.py
"""Rewrite scorer: mechanical scoring and finalization for --fix mode rewrites.

Two-phase invocation:

  Phase 1 — score rewrites mechanically:
    python rewrite_scorer.py --score-rewrites audit.json rewrites_input.json > rewrite_semi.json

  Phase 2 — compose final scores + apply safety gates:
    python rewrite_scorer.py --finalize rewrite_semi.json judgment_patches.json audit.json > rewrites.json
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
import tempfile
from pathlib import Path

if hasattr(sys.stdin, 'reconfigure'):
    sys.stdin.reconfigure(encoding='utf-8')
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')
if hasattr(sys.stderr, 'reconfigure'):
    sys.stderr.reconfigure(encoding='utf-8')

sys.path.insert(0, str(Path(__file__).parent))
import _lib

SCRIPTS_DIR = Path(__file__).parent
PYTHON = sys.executable


def _letter_grade(score: float) -> str:
    if score >= 0.80:
        return "A"
    if score >= 0.65:
        return "B"
    if score >= 0.50:
        return "C"
    if score >= 0.35:
        return "D"
    return "F"


def score_rewrites(audit_path: str, rewrites_input_path: str) -> dict:
    """Phase 1: mechanically score all rewrites through the pipeline."""
    with open(audit_path, encoding="utf-8") as f:
        audit = json.load(f)

    with open(rewrites_input_path, encoding="utf-8") as f:
        rewrites_input = json.load(f)

    if not rewrites_input:
        return {"schema_version": "0.1", "pipeline_version": "0.1.0",
                "project_context": {}, "config": {}, "source_files": [], "rules": []}

    source_files = audit.get("source_files", [{"path": "", "globs": [],
                                                "glob_match_count": 0,
                                                "default_category": "mandate",
                                                "line_count": 1,
                                                "always_loaded": True}])
    if not source_files:
        source_files = [{"path": "", "globs": [], "glob_match_count": 0,
                         "default_category": "mandate", "line_count": 1,
                         "always_loaded": True}]

    rules = []
    for i, rw in enumerate(rewrites_input):
        rules.append({
            "id": rw.get("rule_id", f"RW{i+1:03d}"),
            "file_index": 0,
            "text": rw["suggested_rewrite"],
            "line_start": rw.get("line_start", 1),
            "line_end": rw.get("line_start", 1),
            "category": "mandate",
            "referenced_entities": [],
            "staleness": {"gated": False, "missing_entities": []},
            "factors": {},
        })

    pipeline_input = {
        "schema_version": "0.1",
        "pipeline_version": "0.1.0",
        "project_context": audit.get("project_context", {"stack": []}),
        "config": audit.get("config", {}),
        "source_files": [source_files[0]],
        "rules": rules,
    }

    input_json = json.dumps(pipeline_input, ensure_ascii=False)

    p1 = subprocess.Popen(
        [PYTHON, str(SCRIPTS_DIR / "score_mechanical.py")],
        stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
        text=True, encoding="utf-8",
    )
    p2 = subprocess.Popen(
        [PYTHON, str(SCRIPTS_DIR / "score_semi.py")],
        stdin=p1.stdout, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
        text=True, encoding="utf-8",
    )
    p1.stdout.close()
    p1.stdin.write(input_json)
    p1.stdin.close()

    stdout, stderr2 = p2.communicate(timeout=60)
    p1.wait()

    if p1.returncode != 0:
        print(f"score_mechanical.py failed: {p1.stderr.read()}", file=sys.stderr)
        sys.exit(1)
    if p2.returncode != 0:
        print(f"score_semi.py failed: {stderr2}", file=sys.stderr)
        sys.exit(1)

    result = json.loads(stdout)

    # Fragmentation check: detect rewrites that would split into multiple rules on re-extraction.
    import extract as _extract

    for i, rule in enumerate(result.get("rules", [])):
        if i < len(rewrites_input):
            rw = rewrites_input[i]
            rule["_rewrite_meta"] = {
                "rule_id": rw.get("rule_id"),
                "original_text": rw.get("original_text", ""),
                "file": rw.get("file", ""),
                "line_start": rw.get("line_start", 0),
                "old_score": rw.get("old_score", 0),
                "old_dominant_weakness": rw.get("old_dominant_weakness"),
                "projected_score": rw.get("projected_score"),
            }
            fragments = _extract.would_fragment(rw["suggested_rewrite"])
            if len(fragments) > 1:
                rule["_rewrite_meta"]["would_fragment"] = True
                rule["_rewrite_meta"]["fragment_count"] = len(fragments)
                rule["_rewrite_meta"]["fragments_preview"] = [f[:80] for f in fragments]
                print(
                    f"WARNING: {rw.get('rule_id', f'RW{i+1:03d}')} rewrite would fragment into "
                    f"{len(fragments)} rules when re-extracted. First fragment: "
                    f"{fragments[0][:80]!r}. Revise to use `or` instead of `, and`, "
                    "drop semicolons, or collapse to a single directive.",
                    file=sys.stderr,
                )

    return result


def finalize_rewrites(rewrite_semi_path: str, patches_path: str, audit_path: str) -> list[dict]:
    """Phase 2: compose final scores, apply safety gates, produce rewrites list."""
    with open(rewrite_semi_path, encoding="utf-8") as f:
        rewrite_semi = json.load(f)

    with open(audit_path, encoding="utf-8") as f:
        audit = json.load(f)

    with open(patches_path, encoding="utf-8") as f:
        patches_data = json.load(f)
    patches = patches_data.get("patches", patches_data)
    for rule in rewrite_semi.get("rules", []):
        rule_id = rule.get("id")
        if rule_id not in patches:
            continue
        patch = patches[rule_id]
        if "factors" not in rule:
            rule["factors"] = {}
        for factor_name, factor_data in patch.items():
            if factor_name.endswith("_patch"):
                base_name = factor_name.replace("_patch", "")
                if base_name in rule["factors"] and isinstance(factor_data, dict) and "value" in factor_data:
                    rule["factors"][base_name]["value"] = factor_data["value"]
                    rule["factors"][base_name]["method"] = "judgment_patch"
            elif factor_name in ("F3", "F8"):
                rule["factors"][factor_name] = factor_data

    result = subprocess.run(
        [PYTHON, str(SCRIPTS_DIR / "compose.py")],
        input=json.dumps(rewrite_semi, ensure_ascii=False),
        capture_output=True, text=True, timeout=60, encoding="utf-8",
    )

    if result.returncode != 0:
        print(f"compose.py failed: {result.stderr}", file=sys.stderr)
        sys.exit(1)

    composed = json.loads(result.stdout)
    composed_rules = composed.get("rules", [])

    audit_rules = {r["id"]: r for r in audit.get("rules", [])}

    rewrites = []
    for rule in composed_rules:
        meta = rule.get("_rewrite_meta", {})
        rule_id = meta.get("rule_id", rule.get("id", "?"))
        old_score = meta.get("old_score", 0)
        new_score = rule.get("score", 0)
        old_grade = _letter_grade(old_score)
        new_grade = _letter_grade(new_score)
        projected = meta.get("projected_score")

        # Safety gate 1: Regression — drop rewrites that score lower than original.
        if new_score < old_score:
            continue

        orig_rule = audit_rules.get(rule_id, {})
        orig_factors = orig_rule.get("factors", {})
        new_factors = rule.get("factors", {})
        improvements = {}
        for fn in ("F1", "F2", "F3", "F4", "F7", "F8"):
            old_val = orig_factors.get(fn, {}).get("value")
            new_val = new_factors.get(fn, {}).get("value")
            if old_val is not None and new_val is not None and new_val > old_val:
                improvements[fn] = [round(old_val, 2), round(new_val, 2)]

        # Safety gate 2: Judgment volatility — flag large F3 swings.
        old_f3 = orig_factors.get("F3", {}).get("value")
        new_f3 = new_factors.get("F3", {}).get("value")
        f3_delta = abs((new_f3 or 0) - (old_f3 or 0))
        jv_flagged = f3_delta > 0.20

        # Safety gate 3: Self-verification delta.
        svd = abs(new_score - projected) if projected is not None else 0.0

        rewrites.append({
            "rule_id": rule_id,
            "file": meta.get("file", ""),
            "line_start": meta.get("line_start", 0),
            "original_text": meta.get("original_text", ""),
            "suggested_rewrite": rule.get("text", ""),
            "old_score": round(old_score, 3),
            "new_score": round(new_score, 3),
            "delta": round(new_score - old_score, 3),
            "old_grade": old_grade,
            "new_grade": new_grade,
            "old_dominant_weakness": meta.get("old_dominant_weakness"),
            "new_dominant_weakness": rule.get("dominant_weakness"),
            "factor_improvements": improvements or None,
            "judgment_volatility": {
                "flagged": jv_flagged,
                "f3_delta": round(f3_delta, 2),
                "old_f3": old_f3,
                "new_f3": new_f3,
            },
            "projected_score": projected,
            "self_verification_delta": round(svd, 3),
        })

    return rewrites


def main():
    if len(sys.argv) < 2:
        print("Usage:", file=sys.stderr)
        print("  rewrite_scorer.py --score-rewrites <audit.json> <rewrites_input.json> [--output file]", file=sys.stderr)
        print("  rewrite_scorer.py --finalize <rewrite_semi.json> <patches.json> <audit.json> [--output file]", file=sys.stderr)
        sys.exit(1)

    output_path = None
    filtered = []
    args = sys.argv[1:]
    i = 0
    while i < len(args):
        if args[i] == "--output" and i + 1 < len(args):
            output_path = args[i + 1]
            i += 2
        else:
            filtered.append(args[i])
            i += 1

    if not filtered:
        print("Missing mode argument", file=sys.stderr)
        sys.exit(1)

    mode = filtered[0]
    positional = filtered[1:]

    if mode == "--score-rewrites":
        if len(positional) != 2:
            print("Usage: rewrite_scorer.py --score-rewrites <audit.json> <rewrites_input.json> [--output file]", file=sys.stderr)
            sys.exit(1)
        result = score_rewrites(positional[0], positional[1])
        if output_path:
            with open(output_path, "w", encoding="utf-8") as f:
                json.dump(result, f, indent=2, ensure_ascii=False)
        else:
            _lib.emit(result)

    elif mode == "--finalize":
        if len(positional) != 3:
            print("Usage: rewrite_scorer.py --finalize <rewrite_semi.json> <patches.json> <audit.json> [--output file]", file=sys.stderr)
            sys.exit(1)
        rewrites = finalize_rewrites(positional[0], positional[1], positional[2])
        if output_path:
            with open(output_path, "w", encoding="utf-8") as f:
                json.dump(rewrites, f, indent=2, ensure_ascii=False)
        else:
            json.dump(rewrites, sys.stdout, indent=2, ensure_ascii=False)
            sys.stdout.write("\n")

    else:
        print(f"Unknown mode: {mode}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/rewrite_scorer.js
"use strict";

const path = require("path");
const { spawnSync } = require("child_process");
const _lib = require("./_lib.js");

const SCRIPTS_DIR = __dirname;

function letterGrade(score) {
  if (score >= 0.80) return "A";
  if (score >= 0.65) return "B";
  if (score >= 0.50) return "C";
  if (score >= 0.35) return "D";
  return "F";
}

function runNode(scriptName, inputText) {
  const result = spawnSync(process.execPath, [path.join(SCRIPTS_DIR, scriptName)], {
    input: inputText,
    encoding: "utf-8",
    maxBuffer: 1024 * 1024 * 64,
  });
  return result;
}

function scoreRewrites(auditPath, rewritesInputPath) {
  const audit = _lib.readJson(auditPath);
  const rewritesInput = _lib.readJson(rewritesInputPath);

  if (!rewritesInput || rewritesInput.length === 0) {
    return {
      schema_version: "0.1",
      pipeline_version: "0.1.0",
      project_context: {},
      config: {},
      source_files: [],
      rules: [],
    };
  }

  let sourceFiles = audit.source_files || [
    {
      path: "",
      globs: [],
      glob_match_count: 0,
      default_category: "mandate",
      line_count: 1,
      always_loaded: true,
    },
  ];
  if (!sourceFiles.length) {
    sourceFiles = [
      {
        path: "",
        globs: [],
        glob_match_count: 0,
        default_category: "mandate",
        line_count: 1,
        always_loaded: true,
      },
    ];
  }

  const rules = rewritesInput.map((rw, i) => ({
    id: rw.rule_id || `RW${String(i + 1).padStart(3, "0")}`,
    file_index: 0,
    text: rw.suggested_rewrite,
    line_start: rw.line_start != null ? rw.line_start : 1,
    line_end: rw.line_start != null ? rw.line_start : 1,
    category: "mandate",
    referenced_entities: [],
    staleness: { gated: false, missing_entities: [] },
    factors: {},
  }));

  const pipelineInput = {
    schema_version: "0.1",
    pipeline_version: "0.1.0",
    project_context: audit.project_context || { stack: [] },
    config: audit.config || {},
    source_files: [sourceFiles[0]],
    rules,
  };

  const inputJson = JSON.stringify(pipelineInput);

  const p1 = runNode("score_mechanical.js", inputJson);
  if (p1.status !== 0) {
    process.stderr.write(`score_mechanical.py failed: ${p1.stderr}\n`);
    process.exit(1);
  }

  const p2 = runNode("score_semi.js", p1.stdout);
  if (p2.status !== 0) {
    process.stderr.write(`score_semi.py failed: ${p2.stderr}\n`);
    process.exit(1);
  }

  const result = JSON.parse(p2.stdout);

  const { wouldFragment } = require("./extract.js");

  const resultRules = result.rules || [];
  for (let i = 0; i < resultRules.length; i++) {
    const rule = resultRules[i];
    if (i < rewritesInput.length) {
      const rw = rewritesInput[i];
      rule._rewrite_meta = {
        rule_id: rw.rule_id,
        original_text: rw.original_text || "",
        file: rw.file || "",
        line_start: rw.line_start || 0,
        old_score: rw.old_score || 0,
        old_dominant_weakness: rw.old_dominant_weakness != null ? rw.old_dominant_weakness : null,
        projected_score: rw.projected_score != null ? rw.projected_score : null,
      };
      const fragments = wouldFragment(rw.suggested_rewrite);
      if (fragments.length > 1) {
        rule._rewrite_meta.would_fragment = true;
        rule._rewrite_meta.fragment_count = fragments.length;
        rule._rewrite_meta.fragments_preview = fragments.map((f) => f.slice(0, 80));
        const ruleId = rw.rule_id || `RW${String(i + 1).padStart(3, "0")}`;
        process.stderr.write(
          `WARNING: ${ruleId} rewrite would fragment into ${fragments.length} rules when re-extracted. ` +
            `First fragment: ${JSON.stringify(fragments[0].slice(0, 80))}. Revise to use \`or\` instead of \`, and\`, ` +
            "drop semicolons, or collapse to a single directive.\n"
        );
      }
    }
  }

  return result;
}

function finalizeRewrites(rewriteSemiPath, patchesPath, auditPath) {
  const rewriteSemi = _lib.readJson(rewriteSemiPath);
  const audit = _lib.readJson(auditPath);
  const patchesData = _lib.readJson(patchesPath);
  const patches = patchesData.patches || patchesData;

  for (const rule of rewriteSemi.rules || []) {
    const ruleId = rule.id;
    if (!(ruleId in patches)) continue;
    const patch = patches[ruleId];
    if (!rule.factors) rule.factors = {};
    for (const [factorName, factorData] of Object.entries(patch)) {
      if (factorName.endsWith("_patch")) {
        const baseName = factorName.replace("_patch", "");
        if (
          baseName in rule.factors &&
          factorData &&
          typeof factorData === "object" &&
          "value" in factorData
        ) {
          rule.factors[baseName].value = factorData.value;
          rule.factors[baseName].method = "judgment_patch";
        }
      } else if (factorName === "F3" || factorName === "F8") {
        rule.factors[factorName] = factorData;
      }
    }
  }

  const result = runNode("compose.js", JSON.stringify(rewriteSemi));
  if (result.status !== 0) {
    process.stderr.write(`compose.py failed: ${result.stderr}\n`);
    process.exit(1);
  }

  const composed = JSON.parse(result.stdout);
  const composedRules = composed.rules || [];

  const auditRules = {};
  for (const r of audit.rules || []) auditRules[r.id] = r;

  const rewrites = [];
  for (const rule of composedRules) {
    const meta = rule._rewrite_meta || {};
    const ruleId = meta.rule_id != null ? meta.rule_id : rule.id != null ? rule.id : "?";
    const oldScore = meta.old_score || 0;
    const newScore = rule.score || 0;
    const oldGrade = letterGrade(oldScore);
    const newGrade = letterGrade(newScore);
    const projected = meta.projected_score != null ? meta.projected_score : null;

    // Safety gate 1: Regression — drop rewrites that score lower than original.
    if (newScore < oldScore) continue;

    const origRule = auditRules[ruleId] || {};
    const origFactors = origRule.factors || {};
    const newFactors = rule.factors || {};
    const improvements = {};
    for (const fn of ["F1", "F2", "F3", "F4", "F7", "F8"]) {
      const oldVal = origFactors[fn] ? origFactors[fn].value : undefined;
      const newVal = newFactors[fn] ? newFactors[fn].value : undefined;
      if (oldVal != null && newVal != null && newVal > oldVal) {
        improvements[fn] = [round2(oldVal), round2(newVal)];
      }
    }
    const hasImprovements = Object.keys(improvements).length > 0;

    // Safety gate 2: Judgment volatility — flag large F3 swings.
    const oldF3 = origFactors.F3 ? origFactors.F3.value : undefined;
    const newF3 = newFactors.F3 ? newFactors.F3.value : undefined;
    const f3Delta = Math.abs((newF3 || 0) - (oldF3 || 0));
    const jvFlagged = f3Delta > 0.2;

    // Safety gate 3: Self-verification delta.
    const svd = projected != null ? Math.abs(newScore - projected) : 0.0;

    rewrites.push({
      rule_id: ruleId,
      file: meta.file || "",
      line_start: meta.line_start || 0,
      original_text: meta.original_text || "",
      suggested_rewrite: rule.text || "",
      old_score: round3(oldScore),
      new_score: round3(newScore),
      delta: round3(newScore - oldScore),
      old_grade: oldGrade,
      new_grade: newGrade,
      old_dominant_weakness: meta.old_dominant_weakness != null ? meta.old_dominant_weakness : null,
      new_dominant_weakness: rule.dominant_weakness != null ? rule.dominant_weakness : null,
      factor_improvements: hasImprovements ? improvements : null,
      judgment_volatility: {
        flagged: jvFlagged,
        f3_delta: round2(f3Delta),
        old_f3: oldF3 != null ? oldF3 : null,
        new_f3: newF3 != null ? newF3 : null,
      },
      projected_score: projected,
      self_verification_delta: round3(svd),
    });
  }

  return rewrites;
}

function round2(n) {
  return Math.round(n * 100) / 100;
}

function round3(n) {
  return Math.round(n * 1000) / 1000;
}

function main() {
  const argv = process.argv.slice(2);
  if (argv.length < 1) {
    process.stderr.write("Usage:\n");
    process.stderr.write(
      "  rewrite_scorer.js --score-rewrites <audit.json> <rewrites_input.json> [--output file]\n"
    );
    process.stderr.write(
      "  rewrite_scorer.js --finalize <rewrite_semi.json> <patches.json> <audit.json> [--output file]\n"
    );
    process.exit(1);
  }

  let outputPath = null;
  const filtered = [];
  let i = 0;
  while (i < argv.length) {
    if (argv[i] === "--output" && i + 1 < argv.length) {
      outputPath = argv[i + 1];
      i += 2;
    } else {
      filtered.push(argv[i]);
      i += 1;
    }
  }

  if (!filtered.length) {
    process.stderr.write("Missing mode argument\n");
    process.exit(1);
  }

  const mode = filtered[0];
  const positional = filtered.slice(1);

  if (mode === "--score-rewrites") {
    if (positional.length !== 2) {
      process.stderr.write(
        "Usage: rewrite_scorer.js --score-rewrites <audit.json> <rewrites_input.json> [--output file]\n"
      );
      process.exit(1);
    }
    const result = scoreRewrites(positional[0], positional[1]);
    if (outputPath) {
      _lib.writeJson(outputPath, result);
    } else {
      _lib.emit(result);
    }
  } else if (mode === "--finalize") {
    if (positional.length !== 3) {
      process.stderr.write(
        "Usage: rewrite_scorer.js --finalize <rewrite_semi.json> <patches.json> <audit.json> [--output file]\n"
      );
      process.exit(1);
    }
    const rewrites = finalizeRewrites(positional[0], positional[1], positional[2]);
    if (outputPath) {
      _lib.writeJson(outputPath, rewrites);
    } else {
      process.stdout.write(JSON.stringify(rewrites, null, 2));
      process.stdout.write("\n");
    }
  } else {
    process.stderr.write(`Unknown mode: ${mode}\n`);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

module.exports = { scoreRewrites, finalizeRewrites, letterGrade };
